SMS support callback with received & read & failed notifications

kuzmany opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) <!-- see below -->
| Bug fix?                               | yes/no
| New feature?                           | yes/no
| Deprecations?                          | yes/no
| BC breaks?                             | yes/no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "features" branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
This PR allow any text messages transport add delivered, read or failed notification.
Extend sms_messages table with: deliveredCount, readCount, failedCount column
Extend sms_messages_stat table with is_delivered, is_read, is_failed  column
All this columns are optional and disabled by default. Every transport can enabled it.

This PR is follower of new 2.16 feature Add support for sms replies as a campaign decision https://github.com/mautic/mautic/pull/7449

For DEVS:

Every SMS transport can implements TransportSettingsInterface and implement method enabledSettings with  array of any constant

TransportSettingsInterface::STAT_DELIVERED
TransportSettingsInterface::STAT_READED
TransportSettingsInterface::STAT_FAILED

This add new stats to column and graphs in details

![image](https://user-images.githubusercontent.com/462477/77453503-a2914300-6df7-11ea-8fc5-6503c32f79f0.png)

![image](https://user-images.githubusercontent.com/462477/77453553-b2108c00-6df7-11ea-88e7-4c59858372b9.png)

Usage:

Example Transport

https://gist.github.com/kuzmany/995313c52f82bbb245152404f0c0d53d#file-exampletransport-php-L14

Example of Transport Callback

https://gist.github.com/kuzmany/1635e2684b4728cabb26e0689b3dbf5c

Dispatch Reply or Delivery events

https://gist.github.com/kuzmany/01001d7e49db8e9ddc00bbaa0592e314

#### How to test it:
You have to register a webhook with the service provider. For Twilio, it's done under Messaging for the phone number. The callback is `https://your.mautic.net/sms/twilio/callback`

0. Run migrations first
1. Configure a SMS gateway 
2. Create a campaign with a "send text message" action then a "sends a text message" decision and do something as a result
3. Reply to a SMS 
4. The action attached to the decision should have been executed



#### List backwards compatibility breaks:
1. Refactor app\bundles\SmsBundle\Helper\ReplyHelper.php to app\bundles\SmsBundle\Helper\CallbackHelper.php
2. Remove getMessage() from app\bundles\SmsBundle\Callback\CallbackInterface.php
3. Add getEvent() to app\bundles\SmsBundle\Callback\CallbackInterface.php - should return ReplyEvent() or DeliveryEvent() object

If 3th party addon use that, need change it


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9952"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

kuzmany committed:
Support SMS callback

kuzmany committed:
Merge branch 'features' into sms-callback-support

kuzmany committed:
Merge branch 'features' into sms-callback-support

RCheesley commented:
@kuzmany some [tests failing here](https://github.com/mautic/mautic/pull/9952/checks?check_run_id=2594672139) to look at please!

kuzmany committed:
Fix unit tests

kuzmany committed:
Fix migrations

kuzmany commented:
@RCheesley @npracht could we add this to beta? I will deliver more unit tests and docs during the week. It contains little BC, then Mautic 4 is the best time to deliver it to core.

kuzmany committed:
TransportSettingsTest

RCheesley commented:
@kuzmany may be worth mentioning about applying the migration in the test instructions? When I tried to create a new SMS I am getting a 500 error with a field missing, did not realise there was a migration to be made :)

When I have the PR applied and the migration applied, I am unable to access my SMS section of Mautic due to the following error - the page appears, then within a second or so throws an error:

```
[2021-05-24 19:49:25] mautic.CRITICAL: Uncaught PHP Exception ErrorException: "PHP Warning - Invalid argument supplied for foreach()" at /var/www/html/app/bundles/SmsBundle/Controller/AjaxController.php line 42 {"exception":"[object] (ErrorException(code: 0): PHP Warning - Invalid argument supplied for foreach() at /var/www/html/app/bundles/SmsBundle/Controller/AjaxController.php:42)
[stacktrace]
#0 /var/www/html/app/bundles/SmsBundle/Controller/AjaxController.php(42): Mautic\\CoreBundle\\ErrorHandler\\ErrorHandler->handleError(2, 'PHP Warning - I...', '/var/www/html/a...', 42, Array)
#1 /var/www/html/app/bundles/CoreBundle/Controller/AjaxController.php(136): Mautic\\SmsBundle\\Controller\\AjaxController->getSmsCountStatsAction(Object(Symfony\\Component\\HttpFoundation\\Request), 'sms')
#2 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(158): Mautic\\CoreBundle\\Controller\\AjaxController->executeAjaxAction('getSmsCountStat...', Object(Symfony\\Component\\HttpFoundation\\Request), 'sms')
#3 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#4 /var/www/html/vendor/symfony/framework-bundle/Controller/ControllerTrait.php(96): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#5 /var/www/html/app/bundles/CoreBundle/Controller/AjaxController.php(114): Symfony\\Bundle\\FrameworkBundle\\Controller\\Controller->forward('MauticSmsBundle...', Array)
#6 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(158): Mautic\\CoreBundle\\Controller\\AjaxController->delegateAjaxAction()
#7 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#8 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#9 /var/www/html/app/AppKernel.php(104): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#10 /var/www/html/app/middlewares/CORSMiddleware.php(91): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#11 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(43): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#12 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(61): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#13 /var/www/html/app/middlewares/VersionCheckMiddleware.php(67): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#14 /var/www/html/app/middlewares/TrustMiddleware.php(51): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#15 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#16 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#17 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#18 {main}
"} 
```

RCheesley commented:
When I try to select 'send an SMS' in the campaign, I get the following error:

```
[2021-05-24 19:52:14] mautic.CRITICAL: Uncaught PHP Exception ErrorException: "PHP Notice - Trying to access array offset on value of type null" at /var/www/html/app/bundles/SmsBundle/Form/Type/SmsSendType.php line 84 {"exception":"[object] (ErrorException(code: 0): PHP Notice - Trying to access array offset on value of type null at /var/www/html/app/bundles/SmsBundle/Form/Type/SmsSendType.php:84)
[stacktrace]
#0 /var/www/html/app/bundles/SmsBundle/Form/Type/SmsSendType.php(84): Mautic\\CoreBundle\\ErrorHandler\\ErrorHandler->handleError(8, 'PHP Notice - Tr...', '/var/www/html/a...', 84, Array)
#1 /var/www/html/vendor/symfony/form/ResolvedFormType.php(128): Mautic\\SmsBundle\\Form\\Type\\SmsSendType->buildForm(Object(Symfony\\Component\\Form\\FormBuilder), Array)
#2 /var/www/html/vendor/symfony/form/Extension/DataCollector/Proxy/ResolvedTypeDataCollectorProxy.php(95): Symfony\\Component\\Form\\ResolvedFormType->buildForm(Object(Symfony\\Component\\Form\\FormBuilder), Array)
#3 /var/www/html/vendor/symfony/form/FormFactory.php(80): Symfony\\Component\\Form\\Extension\\DataCollector\\Proxy\\ResolvedTypeDataCollectorProxy->buildForm(Object(Symfony\\Component\\Form\\FormBuilder), Array)
#4 /var/www/html/vendor/symfony/form/FormBuilder.php(94): Symfony\\Component\\Form\\FormFactory->createNamedBuilder('properties', Object(Symfony\\Component\\Form\\Extension\\DataCollector\\Proxy\\ResolvedTypeDataCollectorProxy), NULL, Array)
#5 /var/www/html/vendor/symfony/form/FormBuilder.php(244): Symfony\\Component\\Form\\FormBuilder->create('properties', 'Mautic\\\\SmsBundl...', Array)
#6 /var/www/html/vendor/symfony/form/FormBuilder.php(195): Symfony\\Component\\Form\\FormBuilder->resolveChildren()
#7 /var/www/html/vendor/symfony/form/FormFactory.php(30): Symfony\\Component\\Form\\FormBuilder->getForm()
#8 /var/www/html/app/bundles/CampaignBundle/Controller/EventController.php(89): Symfony\\Component\\Form\\FormFactory->create('Mautic\\\\Campaign...', Array, Array)
#9 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(458): Mautic\\CampaignBundle\\Controller\\EventController->newAction(0, '')
#10 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(158): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('new', 0, 0, '')
#11 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#12 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#13 /var/www/html/app/AppKernel.php(104): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#14 /var/www/html/app/middlewares/CORSMiddleware.php(91): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#15 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(43): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#16 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(61): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#17 /var/www/html/app/middlewares/VersionCheckMiddleware.php(67): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#18 /var/www/html/app/middlewares/TrustMiddleware.php(51): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#19 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#20 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#21 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#22 {main}
"} 

```

kuzmany committed:
Fix Trying to access array offset on value of type null

kuzmany commented:
@RCheesley 
- added migrations instruction
- fixed `Trying to access array offset on value of type null in campaign actions`
- fixed `PHP Warning - Invalid argument supplied for foreach()`

kuzmany committed:
Fix PHP Warning - Invalid argument supplied for foreach()

RCheesley reviewed (dismissed):
Tested this with Twilio, thanks for fixing the issues @kuzmany !

Was able to send an SMS and then reply, based on the response the correct action was taken (in my case I added a tag for the green option and a tag for the red option.

I was not able to see the 'delivered' stats that are demonstrated in the PR description, is that implemented? If so, where?

![screenshot-mautic4 ddev site-2021 05 25-11_54_41](https://user-images.githubusercontent.com/2930593/119486477-2e7e6200-bd50-11eb-9229-831bbc063f3f.png)

![screenshot-mautic4 ddev site-2021 05 25-11_55_02](https://user-images.githubusercontent.com/2930593/119486482-30e0bc00-bd50-11eb-8fea-d4a70f791183.png)

Was able to see the response from the contact in the activity stream, and the text that had been sent:

![screenshot-mautic4 ddev site-2021 05 25-11_55_36](https://user-images.githubusercontent.com/2930593/119486528-3fc76e80-bd50-11eb-9767-ac275f594197.png)

Also tested creating, updating, deleting, cloning of text messages and checked segment & campaign based SMS.

Great work on this @kuzmany - a really nice feature that I am sure will be super helpful.

One small thing, if we are to add this to Mautic 4, please add your BC breaks to the upgrade-4.0.md file so that folk are aware of the deprecations/removals.

Will also need a PR to the dev docs and end user docs to add information about the new feature and how it works.

dadarya0 commented on app/bundles/SmsBundle/Config/config.php:
```diff
@@ -158,8 +175,17 @@
+                    'mautic.sms.model.sms',
+                ],
+            ],
             'mautic.sms.helper.reply' => [
```
I think we can remove this `mautic.sms.helper.reply` from config because we already added `mautic.sms.helper.callback` for Mautic\SmsBundle\Helper\CallbackHelper::class.
What do you think ?

kuzmany replied:
Thank you for suggestions. That config removed and suggestion applied

dadarya0 commented on app/bundles/SmsBundle/Controller/CallbackController.php:
```diff
@@ -13,31 +13,31 @@
+    private $callbackHelper;
 
     /**
      * ReplyController constructor.
```
```suggestion
     * CallbackController constructor.
```

dadarya0 commented on app/bundles/SmsBundle/Entity/Sms.php:
```diff
@@ -143,6 +167,20 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
             ->columnName('sent_count')
             ->build();
 
+        $builder->createField('deliveredCount', 'integer')
```
```suggestion
        $builder->createField('deliveredCount', Types::INTEGER)
```

dadarya0 commented on app/bundles/SmsBundle/Entity/Sms.php:
```diff
@@ -143,6 +167,20 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
+            ->columnName('delivered_count')
+            ->build();
+
+        $builder->createField('readCount', 'integer')
```
```suggestion
        $builder->createField('readCount', Types::INTEGER)
```

dadarya0 commented on app/bundles/SmsBundle/Entity/Sms.php:
```diff
@@ -143,6 +167,20 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
+            ->columnName('read_count')
+            ->build();
+
+        $builder->createField('failedCount', 'integer')
```
```suggestion
        $builder->createField('failedCount', Types::INTEGER)
```

dadarya0 commented on app/bundles/SmsBundle/Entity/Sms.php:
```diff
@@ -143,6 +167,20 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
+            ->columnName('failed_count')
+            ->build();
+
+        $builder->addField('properties', 'json_array');
```
```suggestion
        $builder->addField('properties', Types::JSON);
```

dadarya0 commented on app/bundles/SmsBundle/Entity/Stat.php:
```diff
@@ -118,6 +128,14 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
             ->nullable()
             ->build();
 
+        $builder->createField('isDelivered', 'boolean')
```
```suggestion
        $builder->createField('isDelivered', Types::BOOLEAN)
```

dadarya0 commented on app/bundles/SmsBundle/Entity/Stat.php:
```diff
@@ -118,6 +128,14 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
+            ->columnName('is_delivered')
+            ->build();
+
+        $builder->createField('isRead', 'boolean')
```
```suggestion
        $builder->createField('isRead', Types::BOOLEAN)
```

dadarya0 commented on app/bundles/SmsBundle/Helper/CallbackHelper.php:
```diff
@@ -45,14 +46,24 @@ class ReplyHelper
+    private $smsModel;
+
     /**
      * ReplyHelper constructor.
```
```suggestion
     * CallbackHelper constructor.
```

dadarya0 reviewed (changes requested):
Thank very good job 👍 , I found some minor changes.
I have tested it every thing is working fine but In my case it is not updating `is_delivered` and `is_read` values after successful delivery and read message.

nickveenhof committed:
Merge branch 'features' into sms-callback-support

kuzmany committed:
Merge branch 'features' into sms-callback-support

kuzmany committed:
Update app/bundles/SmsBundle/Controller/CallbackController.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Entity/Sms.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Entity/Sms.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Entity/Sms.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Entity/Sms.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Entity/Stat.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Entity/Stat.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Update app/bundles/SmsBundle/Helper/CallbackHelper.php

Co-authored-by: Saurabh Gupta <48244990+dadarya0@users.noreply.github.com>

kuzmany committed:
Remove mautic.sms.helper.reply

kuzmany committed:
Add Types to use statement

kuzmany committed:
Add Doctrine\DBAL\Types to stat.php

RCheesley committed:
Merge branch 'features' into sms-callback-support

RCheesley commented:
Bumping from the release due to unresolved conflicts.

kuzmany committed:
Add details about deliver event if failed for example

kuzmany committed:
Fix SMS display timeline failed and delivered event

kuzmany committed:
Fix data null exception

kuzmany committed:
Fix stats column without pending count

kuzmany committed:
Merge remote-tracking branch 'upstream/4.x' into sms-callback-support

# Conflicts:
#	app/bundles/LeadBundle/EventListener/SearchSubscriber.php
#	app/bundles/SmsBundle/Helper/CallbackHelper.php

kuzmany committed:
Fix failed sms graph

kuzmany committed:
Fix sms list column headers

kuzmany committed:
Do not process callback If SMS was already removed

kuzmany committed:
Convert sms properties from json_array to json

kuzmany committed:
Add stat_sms_failed_search index migration

kuzmany committed:
Merge branch '4.x' into sms-callback-support

kuzmany commented:
@RCheesley 

> I was not able to see the 'delivered' stats that are demonstrated in the PR description, is that implemented? If so, where?

This is available for developers. We are using delivered stats for our private plugins. 
Twilio need integrate it first,  anybody can pick it up. Twilio should works with features set like before. 
These new columns stats are available for devs and are disabled by default.

RCheesley commented:
Just need to get the PHPSTAN and codestyle fixes sorted! 🚀

RCheesley committed:
Merge branch '4.x' into sms-callback-support

RCheesley commented:
Would be super cool to get this merged @kuzmany if you can rebase to 5.x!

RCheesley commented:
@kuzmany we will have to push this to Mautic 6 if you can't rebase this. Shout when it's done and we can get it tested.

kuzmany commented:
Close favor to M5 version https://github.com/mautic/mautic/pull/13351
